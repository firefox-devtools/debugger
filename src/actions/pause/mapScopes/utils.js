// @flow

import type { Location } from "../../../types";

export function locColumn(loc: Location): number {
  if (typeof loc.column !== "number") {
    // This shouldn't really happen with locations from the AST, but
    // the datatype we are using allows null/undefined column.
    return 0;
  }

  return loc.column;
}
