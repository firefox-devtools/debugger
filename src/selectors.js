// @flow

const sources = require("./reducers/sources");
const pause = require("./reducers/pause");
const breakpoints = require("./reducers/breakpoints");
const eventListeners = require("./reducers/event-listeners");
const ui = require("./reducers/ui");

/**
 * @param object - location
 */

module.exports = {
  getSource: sources.getSource,
  getSourceByURL: sources.getSourceByURL,
  getSourceById: sources.getSourceById,
  getSources: sources.getSources,
  getSourceText: sources.getSourceText,
  getSourceTabs: sources.getSourceTabs,
  getSelectedSource: sources.getSelectedSource,
  getSelectedLocation: sources.getSelectedLocation,
  getPendingSelectedLocation: sources.getPendingSelectedLocation,
  getPrettySource: sources.getPrettySource,

  getBreakpoint: breakpoints.getBreakpoint,
  getBreakpoints: breakpoints.getBreakpoints,
  getBreakpointsForSource: breakpoints.getBreakpointsForSource,
  getBreakpointsDisabled: breakpoints.getBreakpointsDisabled,
  getBreakpointsLoading: breakpoints.getBreakpointsLoading,

  getPause: pause.getPause,
  getChromeScopes: pause.getChromeScopes,
  getLoadedObjects: pause.getLoadedObjects,
  getExpressions: pause.getExpressions,
  getIsWaitingOnBreak: pause.getIsWaitingOnBreak,
  getShouldPauseOnExceptions: pause.getShouldPauseOnExceptions,
  getShouldIgnoreCaughtExceptions: pause.getShouldIgnoreCaughtExceptions,
  getFrames: pause.getFrames,
  getSelectedFrame: pause.getSelectedFrame,

  getEventListeners: eventListeners.getEventListeners,

  getFileSearchState: ui.getFileSearchState,
  getShownSource: ui.getShownSource
};
