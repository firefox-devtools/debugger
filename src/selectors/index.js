/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

export * from "../reducers/expressions";
export * from "../reducers/sources";
export * from "../reducers/pause";
export * from "../reducers/debuggee";
export * from "../reducers/breakpoints";
export * from "../reducers/pending-breakpoints";
export * from "../reducers/ui";
export * from "../reducers/file-search";
export * from "../reducers/ast";
export * from "../reducers/coverage";
export * from "../reducers/project-text-search";
export * from "../reducers/replay";
export * from "../reducers/source-tree";
export { getEventListeners } from "../reducers/event-listeners";
export {
  getQuickOpenEnabled,
  getQuickOpenQuery,
  getQuickOpenType
} from "../reducers/quick-open";

export { getBreakpointAtLocation } from "./breakpointAtLocation";
export { getVisibleBreakpoints } from "./visibleBreakpoints";
export { isSelectedFrameVisible } from "./isSelectedFrameVisible";
export { getCallStackFrames } from "./getCallStackFrames";
export { getVisibleSelectedFrame } from "./visibleSelectedFrame";
export { getRelativeSources } from "./getRelativeSources";
