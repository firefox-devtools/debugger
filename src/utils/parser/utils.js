// @flow

const babylon = require("babylon");
const traverse = require("babel-traverse").default;
const t = require("babel-types");
const { isDevelopment } = require("devtools-config");
const toPairs = require("lodash/toPairs");
const get = require("lodash/get");
const isEmpty = require("lodash/isEmpty");

import type { SourceText, Location, Frame, TokenResolution } from "../../types";

const ASTs = new Map();

const symbolDeclarations = new Map();

type ASTLocation = {
  start: {
    line: number,
    column: number
  },
  end: {
    line: number,
    column: number
  }
};

export type SymbolDeclaration = {
  name: string,
  location: ASTLocation
};

export type FormattedSymbolDeclaration = {
  id: string,
  title: string,
  subtitle: string,
  value: string,
  location: ASTLocation
};

export type SymbolDeclarations = {
  functions: Array<FormattedSymbolDeclaration>,
  variables: Array<FormattedSymbolDeclaration>
};

function _parse(code) {
  return babylon.parse(code, {
    sourceType: "module",

    plugins: ["jsx", "flow"]
  });
}

function parse(text: string) {
  let ast;
  if (!text) {
    return;
  }

  try {
    ast = _parse(text);
  } catch (error) {
    if (isDevelopment()) {
      console.warn("parse failed", text);
    }

    ast = {};
  }

  return ast;
}

function getAst(sourceText: SourceText) {
  if (ASTs.has(sourceText.id)) {
    return ASTs.get(sourceText.id);
  }

  const ast = parse(sourceText.text);
  ASTs.set(sourceText.id, ast);
  return ast;
}

function getFunctionName(path) {
  if (path.node.id) {
    return path.node.id.name;
  }

  const parent = path.parent;
  if (parent.type == "ObjectProperty") {
    return parent.key.name;
  }

  if (parent.type == "ObjectExpression" || path.node.type == "ClassMethod") {
    return path.node.key.name;
  }

  if (parent.type == "VariableDeclarator") {
    return parent.id.name;
  }

  if (parent.type == "AssignmentExpression") {
    if (parent.left.type == "MemberExpression") {
      return parent.left.property.name;
    }

    return parent.left.name;
  }

  return "anonymous";
}

function isFunction(path) {
  return t.isFunction(path) ||
    t.isArrowFunctionExpression(path) ||
    t.isObjectMethod(path) ||
    t.isClassMethod(path);
}

function formatSymbol(symbol: SymbolDeclaration): FormattedSymbolDeclaration {
  return {
    id: `${symbol.name}:${symbol.location.start.line}`,
    title: symbol.name,
    subtitle: `:${symbol.location.start.line}`,
    value: symbol.name,
    location: symbol.location
  };
}

function getVariableNames(path) {
  if (t.isObjectProperty(path) && !isFunction(path.node.value)) {
    return [
      formatSymbol({
        name: path.node.key.name,
        location: path.node.loc
      })
    ];
  }

  if (!path.node.declarations) {
    return path.node.params.map(dec =>
      formatSymbol({
        name: dec.name,
        location: dec.loc
      }));
  }

  return path.node.declarations.map(dec =>
    formatSymbol({
      name: dec.id.name,
      location: dec.loc
    }));
}

function isVariable(path) {
  return t.isVariableDeclaration(path) ||
    (isFunction(path) && path.node.params.length) ||
    (t.isObjectProperty(path) && !isFunction(path.node.value));
}

function getMemberExpression(root) {
  function _getMemberExpression(node, expr) {
    if (t.isMemberExpression(node)) {
      expr = [node.property.name].concat(expr);
      return _getMemberExpression(node.object, expr);
    }

    if (t.isThisExpression(node)) {
      return ["this"].concat(expr);
    }
    return [node.name].concat(expr);
  }

  return _getMemberExpression(root, []);
}

function getSymbols(source: SourceText): SymbolDeclarations {
  if (symbolDeclarations.has(source.id)) {
    const symbols = symbolDeclarations.get(source.id);
    if (symbols) {
      return symbols;
    }
  }

  const ast = getAst(source);
  const symbols = { functions: [], variables: [] };

  traverse(ast, {
    enter(path) {
      if (isVariable(path)) {
        symbols.variables.push(...getVariableNames(path));
      }

      if (isFunction(path)) {
        symbols.functions.push(
          formatSymbol({
            name: getFunctionName(path),
            location: path.node.loc
          })
        );
      }

      if (t.isClassDeclaration(path)) {
        symbols.variables.push(
          formatSymbol({
            name: path.node.id.name,
            location: path.node.loc
          })
        );
      }
    }
  });

  symbolDeclarations.set(source.id, symbols);
  return symbols;
}

function resolveExpression(path, token: string, location: Location): ?Object {
  const node = path.node;
  if (
    t.isMemberExpression(node) &&
    node.property.name === token &&
    nodeContainsLocation({ node, location })
  ) {
    const expr = getMemberExpression(node);
    return {
      value: expr.join("."),
      location: node.loc
    };
  }

  return null;
}

function resolveScope(path, location: Location) {
  const node = path.node;
  if (
    (isFunction(path) || t.isProgram(path)) &&
    nodeContainsLocation({ node, location })
  ) {
    return path;
  }
}

// Resolves a token (at location) in the source to determine if it is in scope
// of the given frame and the expression (if any) to which it belongs
function resolveToken(
  source: SourceText,
  token: string,
  location: Location,
  frame: Frame
): ?TokenResolution {
  const ast = getAst(source);
  const scopes = [];
  let expression = null;
  let inScope = false;

  if (isEmpty(ast)) {
    return;
  }

  traverse(ast, {
    enter(path) {
      let scope = null;

      // if we haven't found an expression yet, determine if the token is part
      // of one
      if (!expression) {
        expression = resolveExpression(path, token, location);
      }

      // determine if the current path is a function or program containing the
      // frame
      scope = resolveScope(path, frame.location);
      if (scope) {
        scopes.unshift(scope);
      }
    }
  });

  // determine if the narrowest scope contains the token's location
  inScope = nodeContainsLocation({ node: scopes[0].node, location });

  return {
    expression,
    inScope
  };
}

function nodeContainsLocation({ node, location }) {
  const { start, end } = node.loc;
  const { line, column } = location;

  return !(start.line > line ||
    (start.line === line && start.column > column) ||
    end.line < line ||
    (end.line === line && end.column < column));
}

function getPathClosestToLocation(source: SourceText, location: Location) {
  const ast = getAst(source);
  let pathClosestToLocation = null;

  traverse(ast, {
    enter(path) {
      if (nodeContainsLocation({ node: path.node, location })) {
        pathClosestToLocation = path;
      }
    }
  });

  return pathClosestToLocation;
}

function getVariablesInScope(source: SourceText, location: Location) {
  const path = getPathClosestToLocation(source, location);
  const bindings = get(path, "scope.bindings", {});

  return toPairs(bindings).map(([name, binding]) => ({
    name,
    references: binding.referencePaths
  }));
}

module.exports = {
  getSymbols,
  getPathClosestToLocation,
  getVariablesInScope,
  resolveToken
};
