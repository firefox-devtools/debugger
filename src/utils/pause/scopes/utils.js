/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import type { Why } from "../../../types";
import type { NamedValue } from "./types";

export function getFramePopVariables(why: Why, path: string): NamedValue[] {
  const vars: Array<NamedValue> = [];

  if (why && why.frameFinished) {
    const frameFinished = why.frameFinished;

    // Always display a `throw` property if present, even if it is falsy.
    if (Object.prototype.hasOwnProperty.call(frameFinished, "throw")) {
      vars.push({
        name: "<exception>",
        path: `${path}/<exception>`,
        contents: { value: frameFinished.throw }
      });
    }

    if (Object.prototype.hasOwnProperty.call(frameFinished, "return")) {
      const returned = frameFinished.return;

      // Do not display undefined. Do display falsy values like 0 and false. The
      // protocol grip for undefined is a JSON object: { type: "undefined" }.
      if (typeof returned !== "object" || returned.type !== "undefined") {
        vars.push({
          name: "<return>",
          path: `${path}/<return>`,
          contents: { value: returned }
        });
      }
    }
  }

  return vars;
}

export function getThisVariable(this_: any, path: string): ?NamedValue {
  if (!this_) {
    return null;
  }

  return {
    name: "<this>",
    path: `${path}/<this>`,
    contents: { value: this_ }
  };
}
