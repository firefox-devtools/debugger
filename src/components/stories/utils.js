const renderComponentFromFixture =
      require("../../test/utils/renderComponentFromFixture");

const { storiesOf } = require("@kadira/storybook");

module.exports = {
  renderComponent: renderComponentFromFixture,
  storiesOf
};
