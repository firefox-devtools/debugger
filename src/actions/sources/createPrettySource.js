import { getFrames, getSource } from "../../selectors";
import { prettyPrint } from "../../workers/pretty-print";
import { updateFrameLocations } from "../../utils/pause";
import { getPrettySourceURL } from "../../utils/source";

export function createPrettySource(sourceId) {
  return async ({ dispatch, getState, sourceMaps }) => {
    const source = getSource(getState(), sourceId).toJS();
    const url = getPrettySourceURL(source.url);
    const id = await sourceMaps.generatedToOriginalId(sourceId, url);

    const { code, mappings } = await prettyPrint({
      source,
      url
    });

    await sourceMaps.applySourceMap(source.id, url, code, mappings);

    let frames = getFrames(getState());
    if (frames) {
      frames = await updateFrameLocations(frames, sourceMaps);
    }

    const prettySource = {
      url,
      id,
      isPrettyPrinted: true,
      text: code,
      contentType: "text/javascript",
      frames,
      loadedState: "loaded"
    };

    dispatch({
      type: "ADD_SOURCE",
      source: prettySource
    });

    return prettySource;
  };
}
