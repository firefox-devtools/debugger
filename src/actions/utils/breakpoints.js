import { getBreakpoint } from "../../selectors";

export async function formatClientBreakpoint(
  clientBreakpoint,
  sourceMaps,
  location
) {
  const clientOriginalLocation = await sourceMaps.getOriginalLocation(
    clientBreakpoint.actualLocation
  );

  // make sure that we are re-adding the same type of breakpoint. Column
  // or line
  const actualLocation = clientOriginalLocation;

  // the generatedLocation might have slid, so now we can adjust it
  const generatedLocation = clientBreakpoint.actualLocation;

  const { id, hitCount } = clientBreakpoint;
  return { id, actualLocation, hitCount, generatedLocation };
}

export function breakpointExists(state, location: Location) {
  const currentBp = getBreakpoint(state, location);
  return currentBp && !currentBp.disabled;
}

export function createBreakpoint(location: Object, overrides: Object = {}) {
  const { condition, disabled, generatedLocation } = overrides;
  const properties = {
    condition: condition || null,
    disabled: disabled || false,
    generatedLocation,
    location
  };

  return properties;
}

export async function getGeneratedLocation(source, sourceMaps, location) {
  if (!sourceMaps.isOriginalId(location.sourceId)) {
    return location;
  }

  return await sourceMaps.getGeneratedLocation(location, source.toJS());
}

export function equalizeLocationColumn(location, referenceLocation) {
  if (referenceLocation.column) {
    return location;
  }
  return { ...location, column: undefined };
}
