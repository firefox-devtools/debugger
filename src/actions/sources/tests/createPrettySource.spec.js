import { actions, createStore, makeSource } from "../../../utils/test-head";
import { createPrettySource } from "../createPrettySource";
import { getSourceByURL } from "../../../selectors";

import { sourceThreadClient } from "../../tests/helpers/threadClient.js";

describe("createPrettySource", () => {
  it("returns a pretty source for a minified file", async () => {
    const { dispatch, getState } = createStore(sourceThreadClient);
    const url = "base.js";
    const source = makeSource(url);
    await dispatch(actions.newSource(source));
    await dispatch(createPrettySource(url));

    const prettyURL = `${source.url}:formatted`;
    const pretty = getSourceByURL(getState(), prettyURL);
    expect(pretty.get("contentType")).toEqual("text/javascript");
    expect(pretty.get("url").includes(prettyURL)).toEqual(true);
    expect(pretty).toMatchSnapshot();
  });
});
