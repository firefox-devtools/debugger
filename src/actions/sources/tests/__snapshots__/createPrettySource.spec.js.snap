// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPrettySource returns a pretty source for a minified file 1`] = `
Immutable.Map {
  url: "http://localhost:8000/examples/base.js:formatted",
  id: "base.js/originalSource-36c718d4bde9a75edb388ff7733efe7f",
  isPrettyPrinted: true,
  text: "undefined
  ",
  contentType: "text/javascript",
  frames: undefined,
  loadedState: "loaded",
}
`;
